File: app/src/main/java/com/radardeal/app/monitoring/VintedQuery.kt
package com.radardeal.app.monitoring

import com.radardeal.app.domain.model.Watch
import java.net.URI
import java.net.URLDecoder
import java.net.URLEncoder

/**
 * Translation between what the user asked for and the URLs RadarDeal actually requests.
 *
 * Two shapes are supported:
 *
 *  * **URL-based watches** — the user pastes a Vinted search URL. Its query string already
 *    encodes every advanced filter (catégorie, taille, état, couleur, marque…), so RadarDeal
 *    forwards those parameters unchanged to the catalogue endpoint that the Vinted web app
 *    itself calls when you scroll that same page.
 *  * **Criteria-based watches** — marque / mot-clé / prix max typed in the app, turned into a
 *    plain text search.
 *
 * This file deliberately uses [java.net.URI] rather than `android.net.Uri` so that the whole
 * mapping is covered by ordinary JVM unit tests.
 */
object VintedQuery {

    /** Marketplaces the app accepts a URL from. Matches "vinted.fr", "www.vinted.co.uk", … */
    private val HOST_PATTERN = Regex("^(?:www\\.)?vinted\\.(?:[a-z]{2,3})(?:\\.[a-z]{2,3})?$")

    /** Vinted's own web front-end asks for this page size. */
    private const val PER_PAGE = 48

    /**
     * Query parameters that are named differently on the public search page and on the
     * catalogue endpoint. Anything not listed here is forwarded under its own name.
     */
    private val PARAM_RENAMES = mapOf(
        "catalog" to "catalog_ids",
        "catalog_id" to "catalog_ids",
        "brand_id" to "brand_ids",
        "size_id" to "size_ids",
        "status_id" to "status_ids",
        "color_id" to "color_ids",
        "material_id" to "material_ids",
    )

    /**
     * Parameters we never forward: they either belong to the HTML page only or would pin the
     * request to a stale result set.
     */
    private val DROPPED_PARAMS = setOf("page", "per_page", "time", "disabled_personalization")

    // --- Public API ------------------------------------------------------------------------

    sealed interface UrlCheck {
        data class Valid(val host: String, val normalizedUrl: String) : UrlCheck
        data class Invalid(val reason: String) : UrlCheck
    }

    /**
     * Validates a pasted URL and reports, in plain French, why it cannot be used. Called from
     * the watch editor as the user types.
     */
    fun checkSearchUrl(raw: String?): UrlCheck {
        val trimmed = raw?.trim().orEmpty()
        if (trimmed.isEmpty()) return UrlCheck.Invalid("Collez une URL de recherche Vinted.")

        val withScheme = if (trimmed.startsWith("http://") || trimmed.startsWith("https://")) {
            trimmed
        } else {
            "https://$trimmed"
        }

        val uri = runCatching { URI(withScheme) }.getOrNull()
            ?: return UrlCheck.Invalid("Cette adresse n'est pas une URL valide.")

        val host = uri.host?.lowercase()
            ?: return UrlCheck.Invalid("Cette adresse n'est pas une URL valide.")

        if (!HOST_PATTERN.matches(host)) {
            return UrlCheck.Invalid("Cette URL ne vient pas de Vinted.")
        }

        val path = uri.path.orEmpty()
        if (path.contains("/items/")) {
            return UrlCheck.Invalid(
                "Ceci est l'adresse d'une annonce, pas d'une recherche. " +
                    "Lancez une recherche sur Vinted puis copiez l'adresse de la page de résultats."
            )
        }

        return UrlCheck.Valid(host = host, normalizedUrl = withScheme)
    }

    /**
     * Endpoint RadarDeal polls, plus the browsable page that "Voir sur Vinted" opens.
     */
    data class Endpoints(
        val host: String,
        val apiUrl: String,
        val browseUrl: String,
    )

    /**
     * Builds the endpoints for [watch]. [fallbackHost] is the marketplace configured in
     * Réglages and is used for criteria-based watches, which carry no host of their own.
     */
    fun endpointsFor(watch: Watch, fallbackHost: String = "www.vinted.fr"): Endpoints {
        val fromUrl = watch.sourceUrl?.let { url ->
            (checkSearchUrl(url) as? UrlCheck.Valid)?.let { valid -> valid to url }
        }

        return if (fromUrl != null) {
            val (valid, originalUrl) = fromUrl
            val params = forwardedParams(valid.normalizedUrl)
            Endpoints(
                host = valid.host,
                apiUrl = buildApiUrl(valid.host, params),
                browseUrl = originalUrl,
            )
        } else {
            val host = fallbackHost.ifBlank { "www.vinted.fr" }
            val params = criteriaParams(watch)
            Endpoints(
                host = host,
                apiUrl = buildApiUrl(host, params),
                browseUrl = buildBrowseUrl(host, params),
            )
        }
    }

    // --- Internals -------------------------------------------------------------------------

    /**
     * Reads the query string of a search page and returns the parameters to forward, with
     * `foo[]=1&foo[]=2` collapsed into `foo=1,2` the way the catalogue endpoint expects.
     */
    internal fun forwardedParams(url: String): LinkedHashMap<String, String> {
        val result = LinkedHashMap<String, String>()
        val query = runCatching { URI(url).rawQuery }.getOrNull().orEmpty()

        val grouped = LinkedHashMap<String, MutableList<String>>()
        for (pair in query.split('&')) {
            if (pair.isBlank()) continue
            val idx = pair.indexOf('=')
            if (idx <= 0) continue
            val rawKey = pair.substring(0, idx)
            val rawValue = pair.substring(idx + 1)
            val key = decode(rawKey).removeSuffix("[]")
            val value = decode(rawValue)
            if (key.isBlank() || value.isBlank()) continue
            if (key in DROPPED_PARAMS) continue
            grouped.getOrPut(PARAM_RENAMES[key] ?: key) { mutableListOf() }.add(value)
        }

        for ((key, values) in grouped) {
            result[key] = values.joinToString(",")
        }
        // Newest first is the only ordering that makes a radar meaningful.
        result["order"] = "newest_first"
        return result
    }

    internal fun criteriaParams(watch: Watch): LinkedHashMap<String, String> {
        val params = LinkedHashMap<String, String>()
        val text = listOfNotNull(
            watch.brand?.trim()?.takeIf { it.isNotEmpty() },
            watch.keyword?.trim()?.takeIf { it.isNotEmpty() },
        ).joinToString(" ")
        if (text.isNotEmpty()) params["search_text"] = text
        watch.minPrice?.let { params["price_from"] = trimNumber(it) }
        watch.maxPrice?.let { params["price_to"] = trimNumber(it) }
        if (watch.minPrice != null || watch.maxPrice != null) params["currency"] = "EUR"
        params["order"] = "newest_first"
        return params
    }

    private fun buildApiUrl(host: String, params: Map<String, String>): String {
        val query = buildString {
            append("page=1&per_page=").append(PER_PAGE)
            for ((key, value) in params) {
                append('&').append(encode(key)).append('=').append(encode(value))
            }
        }
        return "https://$host/api/v2/catalog/items?$query"
    }

    private fun buildBrowseUrl(host: String, params: Map<String, String>): String {
        val query = params.entries.joinToString("&") { (k, v) -> "${encode(k)}=${encode(v)}" }
        return if (query.isEmpty()) "https://$host/catalog" else "https://$host/catalog?$query"
    }

    private fun trimNumber(value: Double): String =
        if (value == value.toLong().toDouble()) value.toLong().toString() else value.toString()

    private fun decode(value: String): String =
        runCatching { URLDecoder.decode(value, "UTF-8") }.getOrDefault(value)

    private fun encode(value: String): String =
        runCatching { URLEncoder.encode(value, "UTF-8") }.getOrDefault(value)
}

File: ios/RadarDealCore/Sources/RadarDealCore/VintedQuery.swift
import Foundation

/// Translation between what the user asked for and the URLs RadarDeal actually requests.
///
/// Two shapes are supported, exactly as on Android:
///
/// * **URL-based watches** — the user pastes a Vinted search URL. Its query string already
///   encodes every advanced filter (catégorie, taille, état, couleur, marque…), so RadarDeal
///   forwards those parameters unchanged to the catalogue endpoint the Vinted web app itself
///   calls when you scroll that same page.
/// * **Criteria-based watches** — marque / mot-clé / prix max typed in the app.
public enum VintedQuery {

    /// Marketplaces a URL is accepted from: "vinted.fr", "www.vinted.co.uk", …
    private static let hostPattern = #"^(?:www\.)?vinted\.(?:[a-z]{2,3})(?:\.[a-z]{2,3})?$"#

    /// Vinted's own web front-end asks for this page size.
    private static let perPage = 48

    /// Parameters named differently on the search page and on the catalogue endpoint.
    private static let renames: [String: String] = [
        "catalog": "catalog_ids",
        "catalog_id": "catalog_ids",
        "brand_id": "brand_ids",
        "size_id": "size_ids",
        "status_id": "status_ids",
        "color_id": "color_ids",
        "material_id": "material_ids",
    ]

    /// Never forwarded: they belong to the HTML page or would pin the request to a stale page.
    private static let dropped: Set<String> = [
        "page", "per_page", "time", "disabled_personalization",
    ]

    // MARK: - URL validation

    public enum URLCheck: Equatable {
        case valid(host: String, normalized: String)
        case invalid(reason: String)
    }

    /// Validates a pasted URL and reports, in plain French, why it cannot be used.
    public static func checkSearchURL(_ raw: String?) -> URLCheck {
        let trimmed = (raw ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            return .invalid(reason: "Collez une URL de recherche Vinted.")
        }

        let withScheme = trimmed.hasPrefix("http://") || trimmed.hasPrefix("https://")
            ? trimmed
            : "https://" + trimmed

        guard
            let components = URLComponents(string: withScheme),
            let host = components.host?.lowercased(),
            !host.isEmpty
        else {
            return .invalid(reason: "Cette adresse n'est pas une URL valide.")
        }

        guard host.range(of: hostPattern, options: .regularExpression) != nil else {
            return .invalid(reason: "Cette URL ne vient pas de Vinted.")
        }

        if components.path.contains("/items/") {
            return .invalid(
                reason: "Ceci est l'adresse d'une annonce, pas d'une recherche. "
                    + "Lancez une recherche sur Vinted puis copiez l'adresse de la page de résultats."
            )
        }

        return .valid(host: host, normalized: withScheme)
    }

    // MARK: - Endpoints

    public struct Endpoints: Equatable {
        public let host: String
        public let apiURL: String
        public let browseURL: String
    }

    /// Builds the endpoints for `watch`. `fallbackHost` is the marketplace configured in
    /// Réglages, used for criteria-based watches which carry no host of their own.
    public static func endpoints(
        for watch: Watch,
        fallbackHost: String = "www.vinted.fr"
    ) -> Endpoints {
        if let source = watch.sourceURL,
           case let .valid(host, normalized) = checkSearchURL(source) {
            let params = forwardedParams(from: normalized)
            return Endpoints(
                host: host,
                apiURL: buildAPIURL(host: host, params: params),
                browseURL: source
            )
        }

        let host = fallbackHost.isEmpty ? "www.vinted.fr" : fallbackHost
        let params = criteriaParams(for: watch)
        return Endpoints(
            host: host,
            apiURL: buildAPIURL(host: host, params: params),
            browseURL: buildBrowseURL(host: host, params: params)
        )
    }

    // MARK: - Parameters

    /// Reads the query string of a search page and returns the parameters to forward, with
    /// `foo[]=1&foo[]=2` collapsed into `foo=1,2` the way the catalogue endpoint expects.
    public static func forwardedParams(from url: String) -> [(String, String)] {
        var grouped: [(key: String, values: [String])] = []

        // The raw query string is split by hand rather than through URLComponents.
        //
        // Foundation is not consistent here: on Darwin `queryItems` percent-decodes values, in
        // swift-corelibs-foundation it does not, and `percentEncodedQueryItems` double-encodes
        // them on Linux ("air%20max" comes back as "air%2520max"). Splitting the string
        // ourselves gives one behaviour on every platform, and the same one the Android
        // implementation has.
        let rawQuery = url.firstIndex(of: "?").map { String(url[url.index(after: $0)...]) } ?? ""

        for pair in rawQuery.split(separator: "&", omittingEmptySubsequences: true) {
            guard let separator = pair.firstIndex(of: "="), separator != pair.startIndex else {
                continue
            }
            let rawKey = String(pair[pair.startIndex..<separator])
            let rawValue = String(pair[pair.index(after: separator)...])

            let value = decode(rawValue)
            guard !value.isEmpty else { continue }

            var key = decode(rawKey)
            if key.hasSuffix("[]") { key = String(key.dropLast(2)) }
            guard !key.isEmpty, !dropped.contains(key) else { continue }
            let mapped = renames[key] ?? key

            if let index = grouped.firstIndex(where: { $0.key == mapped }) {
                grouped[index].values.append(value)
            } else {
                grouped.append((key: mapped, values: [value]))
            }
        }

        var result = grouped.map { ($0.key, $0.values.joined(separator: ",")) }
        // Newest first is the only ordering that makes a radar meaningful.
        result.removeAll { $0.0 == "order" }
        result.append(("order", "newest_first"))
        return result
    }

    public static func criteriaParams(for watch: Watch) -> [(String, String)] {
        var params: [(String, String)] = []

        let text = [watch.brand, watch.keyword]
            .compactMap { $0?.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }
            .joined(separator: " ")
        if !text.isEmpty { params.append(("search_text", text)) }

        if let min = watch.minPrice { params.append(("price_from", trimNumber(min))) }
        if let max = watch.maxPrice { params.append(("price_to", trimNumber(max))) }
        if watch.minPrice != nil || watch.maxPrice != nil {
            params.append(("currency", "EUR"))
        }
        params.append(("order", "newest_first"))
        return params
    }

    // MARK: - Building

    private static func buildAPIURL(host: String, params: [(String, String)]) -> String {
        var query = "page=1&per_page=\(perPage)"
        for (key, value) in params {
            query += "&\(encode(key))=\(encode(value))"
        }
        return "https://\(host)/api/v2/catalog/items?\(query)"
    }

    private static func buildBrowseURL(host: String, params: [(String, String)]) -> String {
        let query = params.map { "\(encode($0.0))=\(encode($0.1))" }.joined(separator: "&")
        return query.isEmpty ? "https://\(host)/catalog" : "https://\(host)/catalog?\(query)"
    }

    /// Reverses query-string encoding: "+" is a space, then percent escapes are resolved.
    private static func decode(_ value: String) -> String {
        let withSpaces = value.replacingOccurrences(of: "+", with: " ")
        return withSpaces.removingPercentEncoding ?? withSpaces
    }

    private static func trimNumber(_ value: Double) -> String {
        value == value.rounded() && abs(value) < 1e15
            ? String(Int64(value))
            : String(value)
    }

    /// Percent-encoding that matches what the Android side produces, including `+` for spaces.
    private static func encode(_ value: String) -> String {
        var allowed = CharacterSet.alphanumerics
        allowed.insert(charactersIn: "-_.*")
        let escaped = value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
        return escaped.replacingOccurrences(of: "%20", with: "+")
    }
}
